### ISSUE TITLE
Custom gradient aggregation methods

### ISSUE THREAD

**Issue by act65** [tensorflow/tensorflow#15760]
I would like a way to apply some custom gradient aggregation ops. Probably the simplest thing to do is just allow `tf.gradients` (and `Optimizer.compute_gradients`) to return un-aggregated gradients so I could work with those?
Anyway, seems like an easy fix? I will do this myself in a month or so (cant now as on holiday), but if someone else wants to do it/has some thoughts, I am interested.

**Comment by yaroslavvb** [comment]
@act65 it's not so straightforward because there's no explicit aggregation step. The "aggregation" happens as part of backprop. You need to create custom backprop to get per-example gradients. See thread https://github.com/tensorflow/tensorflow/issues/4897 for discussion and an example of using khatri-rao product to do this efficiently

**Comment by act65** [comment]
Ah yea, I see the problem. In that case, I think I am after a nicer interface for adding new aggregation methods. Maybe the `aggregation_method` could be passed into 'tf.gradients` as a function/class with the right signature instead?

Regardless, I think I can do what I want (a sketch -- possibly online -- of the grads that returns the principle component) if I just hack in another option to `_AggregatedGrads`.

**Comment by yaroslavvb** [comment]
The classic graph way would be to provide another registration mechanism. IE, using Theano [notation](http://deeplearning.net/software/theano/library/gradient.html) in addition to "right-operator" and "left-operator", each op would support "per-example left-operator", which interprets input nxb matrix as a vectors of  backprops from b independent per-example losses.

However, things seem to be moving in direction of "eager" mode which has much nicer autograd-style wrappers. Maybe @mattjj has some trick up his sleeve on how to accomplish this

**Comment by act65** [comment]
I am not sure it is necessary to add another registration mechanism to allow me to write some custom aggregation ops? It also seems like a lot of work...?
Can we not just think of aggregation as some fn that takes gradients (batch x vectorised parameters) and returns the aggregated grads (1 x vectorised parameters)? So a sum is one instance of this? And we can just drop in some other fn in??

Replace code from [source](https://github.com/tensorflow/tensorflow/blob/r1.4/tensorflow/python/ops/gradients_impl.py).
```python
for i, out_grad in enumerate(out_grads):
    out_grads[i] = math_ops.accumulate_n(out_grad)
```
with 
```python
for i, out_grad in enumerate(out_grads):
    mat = tf.stack(out_grad)
    u, s ,v = tf.svd(mat)
    out_grads[i] = u[:, 0]*s[0]  # the principle component (rather than the sum).
```

But, I feel like I might be missing something? I still dont understand your comment
>  because there's no explicit aggregation step. The "aggregation" happens as part of backprop
I though aggregation over the batch dimension only occurs at the end of the backward propagation of the gradients?!

**Comment by yaroslavvb** [comment]
Aggregation over batch dimension happens at each backprop step. The `_AggregatedGrads` method inside `grad_impl` refer to aggregation over multiple output tensors rather than over examples.

I have a detailed example of backprop derived in [Backprop and systolic arrays](https://medium.com/@yaroslavvb/backprop-and-systolic-arrays-24e925d2050) -- even though the loss in that example is the sum of squared losses, gradients with respect to individual squared losses are never formed, you only have gradient from summed-over loss at any point in time.

If you care about computing some property of per-example gradients rather than full per-example gradient, there are tricks you can do such as @goodfeli 's https://arxiv.org/pdf/1510.01799.pdf

**Comment by act65** [comment]
Ok, I have some reading to do. I will come back to this. Thanks @yaroslavvb

**Comment by hlml** [comment]
@yaroslavvb Hi Yaroslav, I was hoping to get some help implementing the method discussed in https://arxiv.org/pdf/1510.01799.pdf.

@goodfeli suggested the [pseudocode](https://github.com/tensorflow/tensorflow/issues/4897) below:
```
examples = tf.split(batch)
weight_copies = [tf.identity(weights) for x in examples]
output = tf.stack(f(x, w) in zip(examples, weight_copies))
cost = cost_function(output)
per_example_gradients = tf.gradients(cost, weight_copies)
```

But I'm not sure how to implement this still. Would it be possible to show a toy example of this?
